File: app/src/main/java/de/bahnhoefe/deutschlands/bahnhofsfotos/model/ProviderApp.java
package de.bahnhoefe.deutschlands.bahnhofsfotos.model;

import static java.util.stream.Collectors.toList;

import java.util.List;

import lombok.Builder;
import lombok.Value;

@Value
@Builder
public class ProviderApp {

    String countryCode;
    String type;
    String name;
    String url;

    public boolean isAndroid() {
        return "android".equals(type);
    }

    public boolean isWeb() {
        return "web".equals(type);
    }

    public boolean isCompatible() {
        return isAndroid() || isWeb();
    }

    public static boolean hasCompatibleProviderApps(List<ProviderApp> providerApps) {
        return providerApps.stream().anyMatch(ProviderApp::isCompatible);
    }

    public static List<ProviderApp> getCompatibleProviderApps(List<ProviderApp> providerApps) {
        return providerApps.stream()
                .filter(ProviderApp::isCompatible)
                .collect(toList());
    }

}

File: app/src/main/kotlin/de/bahnhoefe/deutschlands/bahnhofsfotos/model/ProviderApp.kt
package de.bahnhoefe.deutschlands.bahnhofsfotos.model

data class ProviderApp(
    val type: String,
    val name: String,
    val url: String,
) {

    val isAndroid: Boolean
        get() = "android" == type
    private val isWeb: Boolean
        get() = "web" == type
    val isCompatible: Boolean
        get() = isAndroid || isWeb

}

fun hasCompatibleProviderApps(providerApps: List<ProviderApp>) =
    providerApps.any { it.isCompatible }

fun getCompatibleProviderApps(providerApps: List<ProviderApp>) =
    providerApps
        .filter { it.isCompatible }
        .toList()
